core/kotlinx-coroutines-core/src/internal/LockFreeTaskQueue.kt

```diff
@@ -87,9 +105,23 @@ internal class LockFreeMPSCQueueCore<E : Any>(private val capacity: Int) {
         _state.loop { state ->
             if (state and (FROZEN_MASK or CLOSED_MASK) != 0L) return state.addFailReason() // cannot add
             state.withState { head, tail ->
-                // there could be one REMOVE element beyond head that we cannot stump up,
+                // If queue is Single-Consumer than there could be one element beyond head that we cannot overwrite,
                 // so we check for full queue with an extra margin of one element
                 if ((tail + 2) and mask == head and mask) return ADD_FROZEN // overfull, so do freeze & copy
+                // If queue is Multi-Consumer then the consumer could still have not cleared element
+                // despite the above check for one free slot.
+                if (!singleConsumer && array[tail and mask] != null) {
```

Comment:
it would be nice if at least one test failed after removal of this block

Reply:
`LockFreeTaskQueueStressTest` now fails when this line is commented out.